ceed-basis.c

```diff
@@ -282,11 +282,95 @@ int CeedBasisView(CeedBasis basis, FILE *stream) {
   return 0;
 }
 
+/// Side to apply Householder matrix
+typedef enum {
+  CEED_LEFT  = 0,
+  CEED_RIGHT = 1,
+} CeedSide;
+
+/// Apply Householder transformation
+/// @param mat        Row-major matrix to be factorized in place
+/// @param v          Householder reflection vector
+/// @param b          Scaling factor
+/// @param n          Number of rows
+/// @param m          Number of columns
+/// @param i          Column to reflect
+int CeedHouseholderApply(CeedScalar *mat, CeedScalar *v, CeedScalar b, CeedInt n,
+                         CeedInt m, CeedInt i, CeedSide side) {
+  // Qtranspose = Hn-1 Hn-2 ... H0
+  //
+  // Hi = |I 0|
+  //      |0 P|
+  //
+  // Left Multiply
+  // |I 0| |A B| = |A   B|
+  // |0 P| |C D|   |PC PD|
+  //
+  // Right Multiply
+  // |A B| |I 0| = |A  PB|
+  // |C D| |0 P|   |C  PD|
+  int j, k;
+  CeedScalar w[n];
+
+  switch (side) {
+  case CEED_LEFT: // Assumes C = 0, in the middle of QR Factorization
+    // Find w = Dtranspose v
+    for (j=i; j<n; j++) { // Row j
+      w[j] = mat[n*i+j];  // v[i] = 1
+      for (k=i+1; k<m; k++) { // Column k
+        w[j] += mat[n*k+j]*v[k];
+      }
+    }
+    // Calculate D - b v wtranspose
+    for (j=i; j<m; j++) { // Row j
+      for (k=i; k<n; k++) { // Column k
+        mat[k+n*j] = mat[k+n*j] - b*v[j]*w[k];
+      }
+    }
+    break;
+  case CEED_RIGHT:
+    // Find w = D v
+    for (j=i; j<n; j++) { // Row j
+      w[j] = mat[i+m*j]; // v[i] = 1
+      for (k=i+1; k<m; k++) { // Column k
+        w[j] += mat[k+m*j]*v[k];
+      }
+    }
+    // Calculate D - b w vtranspose
+    for (j=i; j<n; j++) { // Row j
```

Comment:
@jedbrown I could compact this a bit to account for similarity between lines 318-329 and 332-345, but I think this is what you were getting at.